File: src/app/api/resources.py
from typing import List
from fastapi import APIRouter, Depends, status, UploadFile, File, Query, Request
from sqlalchemy.orm import Session

from app.sql.database import get_db
from app.services.resource_services import ResourceService

from app.models.schemas import (
    ResourceOut,
    AuthorOut,
    CategoryOut,
    KeywordOut,
    ReviewOut,
    ReviewIn
)

router = APIRouter(
    prefix="/resources",
    tags=["Resources"]
)


# ======================================================
# GET /resources
# ======================================================
@router.get("/", response_model=List[ResourceOut])
def list_resources(skip: int = 0, limit: int = 50, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.list_resources(skip, limit)


# ======================================================
# GET /resources/{resource_id}
# ======================================================
@router.get("/{resource_id}", response_model=ResourceOut)
def get_resource(resource_id: int, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.get_resource(resource_id)


# ======================================================
# GET /resources/{resource_id}/authors
# ======================================================
@router.get("/{resource_id}/authors", response_model=List[AuthorOut])
def get_authors(resource_id: int, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.get_authors(resource_id)


# ======================================================
# GET /resources/{resource_id}/categories
# ======================================================
@router.get("/{resource_id}/categories", response_model=List[CategoryOut])
def get_categories(resource_id: int, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.get_categories(resource_id)


# ======================================================
# GET /resources/{resource_id}/keywords
# ======================================================
@router.get("/{resource_id}/keywords", response_model=List[KeywordOut])
def get_keywords(resource_id: int, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.get_keywords(resource_id)


# ======================================================
# GET /resources/{resource_id}/reviews
# ======================================================
@router.get("/{resource_id}/reviews", response_model=List[ReviewOut])
def get_reviews(resource_id: int, db: Session = Depends(get_db)):
    service = ResourceService(db)
    return service.get_reviews(resource_id)


# ======================================================
# POST /resources/{resource_id}/reviews
# ======================================================
@router.post(
    "/{resource_id}/reviews",
    response_model=ReviewOut,
    status_code=status.HTTP_201_CREATED
)
def add_review(
    resource_id: int,
    review_in: ReviewIn,
    db: Session = Depends(get_db)
):
    service = ResourceService(db)
    return service.add_review(
        resource_id=resource_id,
        user_id=review_in.user_id,
        rating=review_in.rating,
        comment=review_in.comment
    )
    
@router.post("/{resource_id}/upload")
async def upload_resource_file(
    resource_id: int,
    file: UploadFile = File(...),
    db: Session = Depends(get_db)
):
    service = ResourceService(db)
    updated_resource = await service.upload_file(resource_id, file)
    return updated_resource

@router.get("/{resource_id}/file")
def download_resource_file(
    resource_id: int,
    user_id: int = Query(None),   # opcional
    request: Request = None,
    db: Session = Depends(get_db),
):
    service = ResourceService(db)

    # Obtener IP y user-agent
    client_ip = request.client.host if request and request.client else None
    user_agent = request.headers.get("user-agent", "unknown") if request else "unknown"

    # Clasificación muy simple de dispositivo
    ua_lower = user_agent.lower()
    if "mobile" in ua_lower:
        device = "mobile"
    elif "tablet" in ua_lower or "ipad" in ua_lower:
        device = "tablet"
    elif "windows" in ua_lower or "macintosh" in ua_lower or "linux" in ua_lower:
        device = "desktop"
    else:
        device = "other"

    metadata = {
        "ip": client_ip,
        "user_agent": user_agent,
        "device": device,
    }

    return service.download_file(resource_id, user_id, metadata)

File: frontend/src/services/resources.ts
import api from './api';
import type { Resource, Author, Category, Keyword, Review, ReviewInput } from '../types';

export const resourcesService = {
  async getAll(skip = 0, limit = 50): Promise<Resource[]> {
    const { data } = await api.get<Resource[]>('/resources/', {
      params: { skip, limit },
    });
    return data;
  },

  async getById(resourceId: number): Promise<Resource> {
    const { data } = await api.get<Resource>(`/resources/${resourceId}`);
    return data;
  },

  async getAuthors(resourceId: number): Promise<Author[]> {
    const { data } = await api.get<Author[]>(`/resources/${resourceId}/authors`);
    return data;
  },

  async getCategories(resourceId: number): Promise<Category[]> {
    const { data } = await api.get<Category[]>(`/resources/${resourceId}/categories`);
    return data;
  },

  async getKeywords(resourceId: number): Promise<Keyword[]> {
    const { data } = await api.get<Keyword[]>(`/resources/${resourceId}/keywords`);
    return data;
  },

  async getReviews(resourceId: number): Promise<Review[]> {
    const { data } = await api.get<Review[]>(`/resources/${resourceId}/reviews`);
    return data;
  },

  async addReview(resourceId: number, review: ReviewInput): Promise<Review> {
    const { data } = await api.post<Review>(`/resources/${resourceId}/reviews`, review);
    return data;
  },

  async uploadFile(resourceId: number, file: File): Promise<Resource> {
    const formData = new FormData();
    formData.append('file', file);
    const { data } = await api.post<Resource>(`/resources/${resourceId}/upload`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return data;
  },

  async downloadFile(resourceId: number, userId?: number): Promise<Blob> {
    const params = userId ? { user_id: userId } : {};
    const { data } = await api.get<Blob>(`/resources/${resourceId}/file`, {
      params,
      responseType: 'blob',
    });
    return data;
  },
};
